x/oracle/keeper/keeper.go

```diff
@@ -110,39 +111,45 @@ func (k Keeper) GetExchangeRateBase(ctx sdk.Context, denom string) (sdk.Dec, err
 	return exchangeRate.Rate.Quo(powerReduction), nil
 }
 
-// SetExchangeRate sets the consensus exchange rate of USD denominated in the
-// denom asset to the store.
-func (k Keeper) SetExchangeRate(ctx sdk.Context, denom string, rate sdk.Dec) {
+// SetExchangeRateWithTimestamp sets the consensus exchange rate of USD denominated in the
+// denom asset to the store with timestamp but with passing timestamp to func
+func (k Keeper) SetExchangeRateWithTimestamp(ctx sdk.Context, denom string, rate sdk.Dec, t time.Time) {
 	key := types.KeyExchangeRate(denom)
-	val := types.ExchangeRate{Rate: rate, Timestamp: ctx.BlockTime()}
-	err := store.SetValue(ctx.KVStore(k.storeKey), key, &val, "exchange_rate")
+	val := types.ExchangeRate{Rate: rate, Timestamp: t}
+	err := store.SetValue[*types.ExchangeRate](ctx.KVStore(k.storeKey), key, &val, "exchange_rate")
 	util.Panic(err)
 }
 
+// SetExchangeRate sets the consensus exchange rate of USD denominated in the
+// denom asset to the store with timestamp
+func (k Keeper) SetExchangeRate(ctx sdk.Context, denom string, rate sdk.Dec) {
+	k.SetExchangeRateWithTimestamp(ctx, denom, rate, ctx.BlockTime())
+}
+
 // SetExchangeRateWithEvent sets an consensus
 // exchange rate to the store with ABCI event
-func (k Keeper) SetExchangeRateWithEvent(ctx sdk.Context, denom string, exchangeRate sdk.Dec) {
-	k.SetExchangeRate(ctx, denom, exchangeRate)
+func (k Keeper) SetExchangeRateWithEvent(ctx sdk.Context, denom string, rate sdk.Dec) {
+	k.SetExchangeRateWithTimestamp(ctx, denom, rate, ctx.BlockTime())
 	sdkutil.Emit(&ctx, &types.EventSetFxRate{
-		Denom: denom, Rate: exchangeRate,
+		Denom: denom, Rate: rate,
 	})
 }
 
 // IterateExchangeRates iterates over all USD rates in the store.
 // TODO: handler should use ExchangeRate type rather than Dec
-func (k Keeper) IterateExchangeRates(ctx sdk.Context, handler func(string, sdk.Dec) bool) {
+func (k Keeper) IterateExchangeRates(ctx sdk.Context, handler func(string, sdk.Dec, time.Time) bool) {
 	store := ctx.KVStore(k.storeKey)
 	iter := sdk.KVStorePrefixIterator(store, types.KeyPrefixExchangeRate)
 	defer iter.Close()
 	prefixLen := len(types.KeyPrefixExchangeRate)
-
 	for ; iter.Valid(); iter.Next() {
 		key := iter.Key()
 		denom := string(key[prefixLen : len(key)-1]) // -1 to remove the null suffix
-		var er types.ExchangeRate
-		k.cdc.MustUnmarshal(iter.Value(), &er)
+		var exgRate types.ExchangeRate
+		err := exgRate.Unmarshal(iter.Value())
```

Comment:
Any reason for the change from `k.cdc.MustUnmarshal` to `exgRate.Unmarshal`?

Reply:
`ExchangeRate` is not a proto message , it is just a normal go struct